onnxruntime/core/optimizer/attention_fusion.cc

```diff
@@ -0,0 +1,711 @@
+}
+
+static NodeArg* ProcessMask(Graph& graph, NodeArg* mask_input, ProviderType provider_type, const logging::Logger& logger) {
+  // Validate mask input shape (batch_size, sequence_length) and data type.
+  // Note that batch_size and sequence_length could be symbolic.
+  const TensorShapeProto* mask_shape = mask_input->Shape();
+  if (mask_shape == nullptr || mask_shape->dim_size() != 2 || mask_input->Type() == nullptr) {
+    DEBUG_LOG("Mask shape is unknown or not 2D, or data type unknown");
+    return nullptr;
+  }
+
+  auto data_type = mask_input->TypeAsProto()->tensor_type().elem_type();
+  if (data_type != ONNX_NAMESPACE::TensorProto_DataType_INT64 &&
+      data_type != ONNX_NAMESPACE::TensorProto_DataType_INT32) {
+    DEBUG_LOG("Mask data type is not int32 or int64");
+    return nullptr;
+  }
+
+  NodeArg* reduce_sum_input = mask_input;
+  if (data_type == ONNX_NAMESPACE::TensorProto_DataType_INT64) {
+    NodeArg& cast_int32 = CastMaskToInt32(graph, mask_input, provider_type);
+    reduce_sum_input = &cast_int32;
+  }
+
+  // Construct shape based on mask input shape. Note that batch_size could be symbolic.
+  TypeProto output_type;
+  output_type.mutable_tensor_type()->set_elem_type(TensorProto_DataType_INT32);
+  auto dim = output_type.mutable_tensor_type()->mutable_shape()->add_dim();
+  *dim = mask_shape->dim(0);
+
+  NodeArg& output = AddMaskReduceSum(graph, reduce_sum_input, output_type, provider_type);
+  return &output;
+}
+
+static NodeArg* GetOrCreateMaskIndex(
+    Graph& graph,
+    NodeArg* mask_input,
+    std::map<std::string, NodeArg*>& mask_index_map,
+    ProviderType provider_type,
+    const logging::Logger& logger) {
+  // Lookup in map, and return the mask index if created.
+  auto search = mask_index_map.find(mask_input->Name());
+  if (search != mask_index_map.end()) {
+    return search->second;
+  }
+
+  NodeArg* output = ProcessMask(graph, mask_input, provider_type, logger);
+  if (nullptr == output) {
+    return nullptr;
+  }
+
+  // Add it to map for lookup later.
+  mask_index_map.insert(std::pair<std::string, NodeArg*>(mask_input->Name(), output));
+  return output;
+}
+
+Status AttentionFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  // A map from mask input arg name to mask index output.
+  std::map<std::string, NodeArg*> mask_index_map;
+
+  int fused_count = 0;
+  for (auto node_index : node_topology_list) {
+    auto* p_node = graph.GetNode(node_index);
+    if (p_node == nullptr)
+      continue;  // we removed the node as part of an earlier fusion
+
+    Node& node = *p_node;
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    if (node.GetOutputEdgesCount() == 4 &&
+        graph_utils::IsSupportedOptypeVersionAndDomain(node, "LayerNormalization", {1}, kOnnxDomain) &&
```

Comment:
>LayerNormalization [](start = 62, length = 18)

LayerNormalization maybe fused into skiplayernorm before running this fusion

Reply:
This transformer is place after LayerNormalization and before SkipLayerNorm fusion.